### ISSUE TITLE
extern method typeParameter not populated in methodInstance.cpp

### ISSUE THREAD

**Issue by hanw** [p4lang/p4c#1856]
@mbudiu-vmw 

I come across a case that need this TODO item to be implemented. Do you remember why this is not implemented before?
https://github.com/p4lang/p4c/blob/master/frontends/p4/methodInstance.cpp#L96

**Comment by hanw** [comment]
For example
` mirror.emit<mirror_metadata_h>(meta.session_id, {meta.srcAddr, meta.lkp_flag2}); `

The `mirror_metadata_h` type information is lost.

**Comment by hanw** [comment]
Tracing down the typeSubstitutionVisitor.cpp pass
```
T/920 -> <Type_Name>(3565) mirror_metadata_h
Removing from generic parameters T/920
Visiting <Type_Var>(399172) T
Replacing T/920 with mirror_metadata_h
Specialized void _<T/920>(in bit<10> session_id, in T/920 hdr)
        into void _<>(in bit<10> session_id, in mirror_metadata_h hdr)
```

The question is why is the type parameter T on the emit method removed?

The extern definition is
```
extern Mirror {
  Mirror();
  void emit<T>(in MirrorId_t session_id, in T hdr);
}

**Comment by mbudiu-vmw** [comment]
After specializing the type parameter the method no longer has a type parameter. Type parameters are only variables. The type variable substitution can be obtained in multiple ways. For example, the type of the second parameter.

**Comment by mbudiu-vmw** [comment]
What do you want to achieve? This part of type checking works quite well, I don't think you should need to change it.

**Comment by jnfoster** [comment]
You've both taken CS 6110! (https://www.cs.cornell.edu/courses/cs6110/2016sp/lectures/lecture30.pdf)

**Comment by hanw** [comment]
I was trying to implement a check in the midend to allow this P4 program,
```
mirror.emit<mirror_metadata_h>(meta.session_id, {meta.srcAddr, meta.lkp_flag2});
```
and disallow this P4 program.
```
mirror.emit(meta.session_id, {meta.srcAddr, meta.lkp_flag2});
```

I tried to use the
`MethodInstance->actualMethodType->typeParameters` to find out if the type has been provided by the programmer and got confused when I found the vector was empty for the first case. But it makes sense now.

Thanks, Nate! Clearly I have put the class material into practice. =P.
